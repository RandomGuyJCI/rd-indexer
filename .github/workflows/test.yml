# This workflow will install Python dependencies, run tests and lint with a single version of Python
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions

name: Python application

on:
  push:
    branches: [ "main" ]
  pull_request:
    branches: [ "main" ]

permissions:
  contents: read

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
      - name: checkout
        uses: actions/checkout@v3
        with:
          lfs: 'true'

      - name: Install poetry
        run: pipx install poetry

      - name: install pypy
        id: setup-pypy
        uses: actions/setup-python@v3
        with:
          python-version: 'pypy-3.8'
          cache: 'poetry'

      - name: install deps 
        run: poetry install --no-interaction
        
      - name: test
        run: poetry run pytest
