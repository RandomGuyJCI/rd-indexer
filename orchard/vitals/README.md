# vitals

This is a script which extracts metadata from a rdzip.

# usage

`poetry run python ./orchard/vitals/vitals.py <path to rdzip>`

