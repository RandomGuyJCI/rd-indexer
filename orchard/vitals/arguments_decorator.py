import functools


def with_arguments(*argument_names):
    """
    A decorator that we put on facets to let facets declare what arguments they actually use.
    The decorator fetches the required values from the dict and gives them to the function in order.
    """

    def decorator(func):
        @functools.wraps(func)
        def wrapper(obj):
            args = [obj[k] for k in argument_names]
            return func(*args)

        return wrapper

    return decorator
