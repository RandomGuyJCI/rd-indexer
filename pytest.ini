[pytest]
addopts = --random-order --random-order-bucket=package