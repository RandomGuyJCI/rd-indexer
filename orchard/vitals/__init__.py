from orchard.vitals.vitals import main as analyze