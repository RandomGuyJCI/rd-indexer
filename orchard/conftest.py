import httpx
import pytest
import debugpy
from playhouse.sqlite_ext import SqliteExtDatabase
import tempfile

from unittest.mock import patch

@pytest.fixture
def empty_db():
    # Use an actual file, as sqlite will terminate in-memory databases early if the connection is closed.
    with tempfile.NamedTemporaryFile() as f:
        db = SqliteExtDatabase(f.name)
        yield db
        db.close()

@pytest.fixture
def debug():
    debugpy.listen(5678)
    debugpy.wait_for_client()  # blocks execution until client is attached


@pytest.fixture(autouse=True)
def block_httpx(httpx_mock):
    yield

@pytest.fixture
def assert_all_responses_were_requested(autouse=True) -> bool:
    return False