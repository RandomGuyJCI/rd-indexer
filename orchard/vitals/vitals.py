import sys
import zipfile
from typing import BinaryIO
from orchard.parse import parse
from orchard.vitals.facets.author_facet import author_facet
from orchard.vitals.facets.beat_types_facet import beat_type_facet
from orchard.vitals.facets.bpm_facet import bpm_facet
from orchard.vitals.facets.difficulty_facet import difficulty_facet
from orchard.vitals.facets.icon_facet import icon_facet
from orchard.vitals.facets.id_facet import id_facet
from orchard.vitals.facets.key_facet import make_key_facet
from orchard.vitals.facets.player_facet import player_facet
from orchard.vitals.facets.tags_facet import tags_facet
from orchard.vitals.facets.thumbnail_facet import thumbnail_facet
from orchard.vitals.facets.updated_facet import updated_facet


class VitalsException(Exception):
    pass


def main(f: BinaryIO):
    facets = {
        "id":                               id_facet,
        "artist":                           make_key_facet(["settings", "artist"]),
        "song":                             make_key_facet(["settings", "song"]),
        "seizure_warning":                  make_key_facet(["settings", "seizureWarning"], True),
        "description":                      make_key_facet(["settings", "description"]),
        "hue":                              make_key_facet(["settings", "songNameHue"], 0.0),
        "authors":                          author_facet,
        ("max_bpm", "min_bpm"):             bpm_facet,
        "difficulty":                       difficulty_facet,
        ("single_player", "two_player"):    player_facet,
        "last_updated":                     updated_facet,
        "tags":                             tags_facet,
        ("image", "thumb"):                 thumbnail_facet,
        "icon":                             icon_facet,
        (
            "has_classics",
            "has_oneshots",
            "has_squareshots",
            "has_freezeshots",
            "has_freetimes",
            "has_holds"
        ):                                  beat_type_facet
    }

    try:
        with zipfile.ZipFile(f) as z:
            with z.open('main.rdlevel', 'r') as rdlevel:
                text = rdlevel.read().decode('utf-8-sig')
                parsed = parse(text)

                final = {}
                for key, func in facets.items():
                    try:
                        result = func(parsed, z)
                        if isinstance(key, tuple):
                            # multiple keys with (expected) multiple values that map to the keys.
                            for k, v in zip(key, result):
                                final[k] = v
                        else:
                            final[key] = result
                    except Exception as e:
                        raise VitalsException(f"vitals: An unhandled error occured in a facet: {e}")

                return final

    except zipfile.BadZipFile:
        raise VitalsException("vitals: this is not a zip file, or we couldn't decode it for some reason.")
    except KeyError:
        raise VitalsException("vitals: there is no main.rdlevel in the zip.")

if __name__ == "__main__":
    try:
        file_name = sys.argv[1]
        with open(file_name, 'rb') as f:
            vit = main(f)
    except IndexError:
        print("Did you pass a file name in?")