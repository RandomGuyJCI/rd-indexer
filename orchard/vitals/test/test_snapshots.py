import pytest
from syrupy.extensions.single_file import SingleFileSnapshotExtension
from orchard.vitals.test.conftest import pt_download_file

from orchard.vitals.vitals import main
from syrupy.filters import props

# for vitals, we just test with snapshots.
# all levels used for snapshots are in the "fixtures" directory.

# snapshots don't necessarily show correctness of the data, just that it hasn't changed.
# so I have to remember to check the data added to the snapshot when I add new things to vitals.

EXCLUDE_FROM_SNAPSHOT = [
    "image",
    "thumb",
    "icon"
]

@pytest.mark.parametrize(["url", "sha1"], [
    ("https://codex.rhythm.cafe/know-you-2zRJRCjPqJx.rdzip", "f2ad0f76bbfedf96bec3451eb9ead014accb3047"),
    ("https://codex.rhythm.cafe/rgb-7krFijAGF4D.rdzip", "88f48cdbd6a63773fcf44a38f5ed1ece00f399be"),
    ("https://codex.rhythm.cafe/pull-EpgEYcGYPnU.rdzip", "bdbd0b30d6ea29f8a6d2e8602b8829508dcbfc02"),
    ("https://codex.rhythm.cafe/swing-c2QGYcw8GLq.rdzip", "933bbbbd9c1d79a8c3d2946e5748548a414488ea"),
    ("https://codex.rhythm.cafe/sol-e-ma-dVMtiNt856v.rdzip", "f909b918e876d72fe8c3d385ba03940f85b3b906"),
    ("https://codex.rhythm.cafe/ennui-og-3VFE7LSmzSo.rdzip", "250da27a5bff4285dcbf22f13c9c287f7eaf30d7"),
    ("https://codex.rhythm.cafe/one-forg-NqVywpSrZs1.rdzip", "a75d03f274c9032a6f75fa32ac57005615c04b47"),
    ("https://codex.rhythm.cafe/one-uncu-eqSkB1ZX7Xz.rdzip", "1bf690214be382e78daa98a5c7e73b79a59cf610"),
    ("https://cdn.discordapp.com/attachments/439363565007142912/1040242016593920030/Samario_Making_A_Mistake_-_One_Uncued_Tresillo2.rdzip", "495bd9d1b759db53e600b3986df55bdf3a9c3653"),
    ("https://cdn.discordapp.com/attachments/439363565007142912/1040248915414503464/auburnsummer_-_Triplet_Testaaaaa1.rdzip", "460807d4b87deb9d94ab31369985ba96b5c95291"),
    ("https://codex.rhythm.cafe/yi-xiao-Hg3afme47nr.rdzip", "82dfb357c88904400fbf7eebc2ff2d1820ad9f8f"),
    ("https://codex.rhythm.cafe/banas-QbMoyaG2nfK.rdzip", "862f1ad3d2f6ae0580055700d79b8ae30c8771b2"),
    ("https://codex.rhythm.cafe/jimmy-s-Ccby1RK3XJK.rdzip", "5bb2d7298237d15b44fbb0a9ca963c00ecb58866"),
])
def test_snapshot(snapshot, pt_download_file, url, sha1):
    with pt_download_file(url, sha1) as f:
        f.seek(0)
        actual = main(f)
        assert actual == snapshot(exclude=props(*EXCLUDE_FROM_SNAPSHOT))
