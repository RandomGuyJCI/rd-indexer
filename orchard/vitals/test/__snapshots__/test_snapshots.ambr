# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\Andrew_Huang_-_Know_You.rdzip]
  dict({
    'artist': 'Andrew Huang',
    'artist_tokens': list([
      'Andrew Huang',
    ]),
    'authors': list([
      'lugi',
    ]),
    'description': 'Two lonely stars, wandering in the dark...',
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 42,
      }),
    ]),
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.47,
    'id': 'know-you-2zRJRCjPqJx',
    'last_updated': datetime.datetime(2020, 6, 24, 15, 6, 48),
    'max_bpm': 78.5,
    'min_bpm': 78.5,
    'rdlevel_sha1': 'f67aa02893fdb5ccfad0e445486bd6595801ecdc',
    'seizure_warning': False,
    'sha1': 'f2ad0f76bbfedf96bec3451eb9ead014accb3047',
    'single_player': True,
    'song': 'Know You',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 8,
      }),
    ]),
    'tags': list([
      'slow',
      'pop',
    ]),
    'two_player': True,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\Hot_Freaks_-_Pull.rdzip]
  dict({
    'artist': 'Hot Freaks',
    'artist_tokens': list([
      'Hot Freaks',
    ]),
    'authors': list([
      'Labyrinth',
    ]),
    'description': "It's not something easy to admit.",
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 33,
      }),
    ]),
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': False,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.88,
    'id': 'pull-EpgEYcGYPnU',
    'last_updated': datetime.datetime(2022, 6, 16, 0, 8, 26),
    'max_bpm': 89,
    'min_bpm': 88,
    'rdlevel_sha1': '9f576dd7ed2430c85d3fed185f259eb13b861669',
    'seizure_warning': False,
    'sha1': '8110b16e256491dcef0baf38bf155609afa60f55',
    'single_player': True,
    'song': 'Pull',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 4,
      }),
    ]),
    'tags': list([
      '1p',
      '2p',
      'gimmick',
    ]),
    'two_player': True,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\JustMe_-_Swing.rdzip]
  dict({
    'artist': 'JustMe',
    'artist_tokens': list([
      'JustMe',
    ]),
    'authors': list([
      'Xeno',
    ]),
    'description': "Printin' twang.",
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 15,
      }),
    ]),
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.08,
    'id': 'swing-c2QGYcw8GLq',
    'last_updated': datetime.datetime(2020, 12, 1, 18, 30, 20),
    'max_bpm': 130,
    'min_bpm': 130,
    'rdlevel_sha1': 'a8fd55c9e69a331a72f2c10a396ebde920b91315',
    'seizure_warning': False,
    'sha1': '3536805b7678006664d4f45856c4ce8959eb4a15',
    'single_player': True,
    'song': 'Swing',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 5,
      }),
    ]),
    'tags': list([
      'Printing',
      'CMYK',
      'Swing',
      'Bounce',
      'syncopation',
      'gimmick',
      '1p',
      '2p',
      'XenOnly',
    ]),
    'two_player': True,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\Pseuu_-_ennui.rdzip]
  dict({
    'artist': 'Pseuu',
    'artist_tokens': list([
      'Pseuu',
    ]),
    'authors': list([
      'Pseuu',
    ]),
    'description': '''
      uh... no comment. level made in about 3-4 working hours
      
      - tricky oneshot patterns
      - skipshots
      - holds
    ''',
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 102,
      }),
    ]),
    'difficulty': 2,
    'has_classics': False,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': True,
    'has_oneshots': True,
    'has_skipshots': True,
    'has_squareshots': False,
    'has_window_dance': True,
    'hue': 0.42,
    'id': 'ennui-og-3VFE7LSmzSo',
    'last_updated': datetime.datetime(2022, 1, 3, 23, 19, 26),
    'max_bpm': 126,
    'min_bpm': 126,
    'rdlevel_sha1': 'c3435bd5e2362805b1ff15629757693ad985083c',
    'seizure_warning': False,
    'sha1': 'fd9bf6a51ce68fd87d3d6b269c792386f86c4c3b',
    'single_player': True,
    'song': 'ennui.ogg',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 9,
      }),
    ]),
    'tags': list([
      'chiptune',
      'short',
      '2p',
      'oneshots',
      'skipshots',
      'holds',
    ]),
    'two_player': True,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\Quree_-_One_Forgotten_Night.rdzip]
  dict({
    'artist': 'Quree',
    'artist_tokens': list([
      'Quree',
    ]),
    'authors': list([
      'BluPen',
    ]),
    'description': "Patient experiencing sudden heart strain again. Solve the patient's concerns.",
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 77,
      }),
    ]),
    'difficulty': 2,
    'has_classics': True,
    'has_freetimes': False,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': True,
    'hue': 0.1,
    'id': 'one-forg-LJfg3wQANR7',
    'last_updated': datetime.datetime(2021, 5, 20, 21, 9, 22),
    'max_bpm': 173,
    'min_bpm': 100,
    'rdlevel_sha1': '9c1236f7939cc820b294dabe8904679701e1011b',
    'seizure_warning': False,
    'sha1': '365fc035a0024c9fe69d51e938fa3d6952b302db',
    'single_player': True,
    'song': 'One Forgotten Night',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 19,
      }),
    ]),
    'tags': list([
      'quree',
      'window dance',
    ]),
    'two_player': True,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\Samario (Making A Mistake) - One Uncued Tresillo.rdzip]
  dict({
    'artist': 'Samario (Making A Mistake)',
    'artist_tokens': list([
      'Samario',
    ]),
    'authors': list([
      'Samario (Making A Mistake)',
    ]),
    'description': '''
      One Uncued Tresillo
      (song edited from [...]Machinegun Psystyle[...], by Camellia)
    ''',
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 81,
      }),
    ]),
    'difficulty': 0,
    'has_classics': False,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': False,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0,
    'id': 'one-uncu-i64TVVfB93P',
    'last_updated': datetime.datetime(2022, 6, 27, 21, 27, 12),
    'max_bpm': 70,
    'min_bpm': 40,
    'rdlevel_sha1': '45b791ffa43c995db620315f968047c6d9c7bb21',
    'seizure_warning': False,
    'sha1': '03a597ff04c86c751e6b12d79d6308b18d1c68f3',
    'single_player': True,
    'song': 'One Uncued Tresillo',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 19,
      }),
    ]),
    'tags': list([
      'uncued tresillo',
      'samario please stop doing this',
    ]),
    'two_player': True,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\Samario_Making_A_Mistake_-_One_Uncued_Tresillo orig.rdzip]
  dict({
    'artist': 'Samario (Making A Mistake)',
    'artist_tokens': list([
      'Samario (Making A Mistake)',
    ]),
    'authors': list([
      'Samario (Making A Mistake)',
    ]),
    'description': '''
      One Uncued Tresillo
      (song edited from [...]Machinegun Psystyle[...], by Camellia)
    ''',
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 81,
      }),
    ]),
    'difficulty': 0,
    'has_classics': False,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': False,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0,
    'id': 'one-uncu-eqSkB1ZX7Xz',
    'last_updated': datetime.datetime(2021, 1, 2, 3, 52, 10),
    'max_bpm': 160,
    'min_bpm': 160,
    'rdlevel_sha1': '6ff955c37225bb5ed9282b3f813cb16dd7504fd3',
    'seizure_warning': False,
    'sha1': '7546233f165a8842b335404d8df3d977888fce50',
    'single_player': True,
    'song': 'One Uncued Tresillo',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 19,
      }),
    ]),
    'tags': list([
      'uncued tresillo',
      'samario please stop doing this',
    ]),
    'two_player': True,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\auburnsummer - Triplet Testaaaaa.rdzip]
  dict({
    'artist': 'auburnsummer',
    'artist_tokens': list([
      'auburnsummer',
    ]),
    'authors': list([
      'auburnsummer',
      ' Xeno',
    ]),
    'description': 'A two-handed level. The eloquent blend between thirds and halves.',
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 65,
      }),
    ]),
    'difficulty': 2,
    'has_classics': True,
    'has_freetimes': False,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': True,
    'has_window_dance': False,
    'hue': 0.66,
    'id': 'triplet-LrxHzPMLamy',
    'last_updated': datetime.datetime(2022, 6, 27, 23, 25, 28),
    'max_bpm': 110,
    'min_bpm': 110,
    'rdlevel_sha1': '7e3f5d28f8865e3e8bc3982b2f72e5d6a9dbddf0',
    'seizure_warning': False,
    'sha1': '4de04dc0c7e6ba30438368cacbd8a24678666841',
    'single_player': True,
    'song': 'Triplet Test',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 12,
      }),
    ]),
    'tags': list([
      'Piano',
      '2p',
      'Two-handed',
    ]),
    'two_player': False,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\color-ff-XsSFVSBp2hH.rdzip]
  dict({
    'artist': '遊助',
    'artist_tokens': list([
      '遊助',
    ]),
    'authors': list([
      '来因洛特',
    ]),
    'description': '这次的这个谱我真的花了很长时间，大概有......差不多50个小时吧。歌词因为是CC，只能全用平假名了，看起来麻烦也抱歉了...希望这次能取得比上次更好的结果吧。来来回回也看了很多次，感觉想不到有什么能加的了，就这样交了吧。也希望你也能像歌词一样开开心心的，活着本身没有意义，但就是为了知道是为了什么我们才会活着呀。（录的时候请把轨道分成八行）',
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 172,
      }),
    ]),
    'difficulty': 0,
    'has_classics': True,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': False,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.17,
    'id': 'color-ff-XsSFVSBp2hH',
    'last_updated': datetime.datetime(2022, 8, 25, 22, 29, 56),
    'max_bpm': 138,
    'min_bpm': 138,
    'rdlevel_sha1': 'a04ccd13fb8ab685880d4223737d943c0e494376',
    'seizure_warning': True,
    'sha1': '72b6e91fe98ff2db4b5f6ff08d75eb0fda71eca1',
    'single_player': True,
    'song': '''
      一笑懸命
    ''',
    'song_ct': list([
      dict({
        'color': '#FF7C8A',
        'len': 1,
      }),
      dict({
        'color': '#FECD66',
        'len': 1,
      }),
      dict({
        'color': '#C3F3FD',
        'len': 1,
      }),
      dict({
        'color': '#FFA3D6',
        'len': 1,
      }),
      dict({
        'color': 'default',
        'len': 1,
      }),
    ]),
    'tags': list([
      'easy',
      'CNcompo 1',
      'relaxing',
      'lyrics',
    ]),
    'two_player': False,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\placeboing_-_BANAS.rdzip]
  dict({
    'artist': 'placeboing',
    'artist_tokens': list([
      'placeboing',
    ]),
    'authors': list([
      "April's Only Fool",
    ]),
    'description': '''
      This level works under the assumption that you know how 8beat, 6beat, and Temporal Shift works.
      
      Hope you all have a wonderful April Fools Day!
      
      Hollaback Girl by Gwen Stefani x Donkey Kong Country 2 OST
      
      Banana CC by Pyre
      Banana Tile by Het
    ''',
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 241,
      }),
    ]),
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': True,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': True,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.17,
    'id': 'banas-QbMoyaG2nfK',
    'last_updated': datetime.datetime(2022, 4, 9, 16, 40, 34),
    'max_bpm': 147,
    'min_bpm': 143.5,
    'rdlevel_sha1': '77bc7d7ea5203e1befdf0bc24ae2fad0e5b0346e',
    'seizure_warning': False,
    'sha1': '0c55a1c001072adcfbc9c1946dbee439b091f00f',
    'single_player': True,
    'song': 'BANAS',
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 5,
      }),
    ]),
    'tags': list([
      'April Fools',
      'banana',
      'banas',
      'temporal shift',
      '8beat',
      '6beat',
    ]),
    'two_player': False,
  })
# ---
# name: test_snapshot[C:/Users/auburn/PycharmProjects/rd-indexer/orchard/vitals/test/fixtures\\toonlink_Kenichi_Nishikami_Takeru_Kanazaki_-_Jimmys_Aesthetics.rdzip]
  dict({
    'artist': 'toonlink, Kenichi Nishikami, Takeru Kanazaki',
    'artist_tokens': list([
      'toonlink',
      'Kenichi Nishikami',
      'Takeru Kanazaki',
    ]),
    'authors': list([
      'Szprycha',
    ]),
    'description': 'A neat song from WarioWare DIY',
    'description_ct': list([
      dict({
        'color': 'default',
        'len': 30,
      }),
    ]),
    'difficulty': 1,
    'has_classics': True,
    'has_freetimes': False,
    'has_freezeshots': False,
    'has_holds': False,
    'has_oneshots': True,
    'has_skipshots': False,
    'has_squareshots': False,
    'has_window_dance': False,
    'hue': 0.19,
    'id': 'jimmy-s-Ccby1RK3XJK',
    'last_updated': datetime.datetime(2020, 6, 20, 23, 9, 28),
    'max_bpm': 125,
    'min_bpm': 125,
    'rdlevel_sha1': 'd61958f09545cf365b4d0c23d12ace8942bc20ec',
    'seizure_warning': False,
    'sha1': '857f166d1a8d30a2bd810de4113eb19260acab5e',
    'single_player': True,
    'song': "Jimmy's Aesthetics",
    'song_ct': list([
      dict({
        'color': 'default',
        'len': 18,
      }),
    ]),
    'tags': list([
      'siivagunner',
      'nintendo',
      'warioware',
      'body rock',
    ]),
    'two_player': False,
  })
# ---
